AppController/djinn.rb

```diff
@@ -5648,37 +5658,48 @@ def try_to_scale_up(app_name, delta_appservers)
 
 
   # Try to remove an AppServer for the specified application, ensuring
-  # that a minimum number of AppServers is always kept.
+  # that a minimum number of AppServers is always kept. We remove
+  # AppServers from the 'latest' appengine node.
   #
   # Args:
   #   app_name: A String containing the application ID.
   #   delta_appservers: The desired number of AppServers to remove.
   # Returns:
   #   A boolean indicating if an AppServer was removed.
-  def try_to_scale_down(app_name, _delta_appservers)
+  def try_to_scale_down(app_name, delta_appservers)
     # See how many AppServers are running on each machine. We cannot scale
     # if we already are at the requested minimum.
     min = @app_info_map[app_name]['min_appengines']
     min = Integer(@options['appengine']) if min.nil?
     if @app_info_map[app_name]['appengine'].length <= min
-      Djinn.log_debug("We are already at the minimum number of AppServers for " +
-        "#{app_name}: requesting to remove node.")
-
-      # If we're this far, nobody can scale down, so try to remove a node instead.
-      ZKInterface.request_scale_down_for_app(app_name, my_node.private_ip)
+      Djinn.log_debug("We are already at the minimum number of AppServers for #{app_name}.")
       return false
     end
 
-    # We pick a randon appengine that run the application.  Smarter
-    # algorithms could be implemented, but without clear directives (ie
-    # decide on cpu, or memory, or number of CPU available, or avg load
-    # etc...) any static strategy is flawed, so we go for simplicity.
-    scapegoat = @app_info_map[app_name]['appengine'].sample()
-    appserver_to_use, port = scapegoat.split(":")
-    Djinn.log_info("Removing an AppServer from #{appserver_to_use} for #{app_name}")
+    # Make sure we leave at least the minimum number of AppServers
+    # running.
+    if delta_appservers > @app_info_map[app_name]['appengine'].length + min
+      num_to_remove = @app_info_map[app_name]['appengine'].length - min
+    else
+      num_to_remove = delta_appservers
+    end
+
+    # Let's pick the latest appengine node hosting the application and
+    # remove the AppServer there, so we can try to reclaim it once it's
+    # unloaded.
+    get_all_appengine_nodes.reverse { |node_ip|
+      @app_info_map[app_name]['appengine'].each { |location|
+        host, port = location.split(":")
+        if host == node_ip
+          @app_info_map[app_name]['appengine'].delete(location)
+          @last_decision[app_name] = Time.now.to_i
```

Comment:
to be safe, I tried to add a lock around any changes of @app_info_map since it's a global structure and it's used in threads extensively.